Add documentation on how to use MySQL instead of PostgreSQL. (#567)
dunglas/symfony-docker: docs/mysql.md

# Using MySQL

> The Docker configuration in this repository defaults to using PostgreSQL.
> If you prefer to work with MySQL, some adjustments need to be made.

Reviewer: ```suggestion
The Docker configuration of this repository is extensible thanks to Flex recipes. By default, the recipe installs PostgreSQL.
If you prefer to work with MySQL, follow these steps:
```
WDYT ?

> First, install the doctrine/orm pack as described: `make composer c='req symfony/orm-pack`

Reviewer 2: We don't use make in the default template, please use compose directly.

> ## Docker configuration

Reviewer 2: ```suggestion
## Docker Configuration
```

> Change the database image to use mysql instead of postgresql in `compose.yaml`:

Reviewer 2: ```suggestion
Change the database image to use MySQL instead of PostgreSQL in `compose.yaml`:
```

Author: yes, my intention was to mention the image names. I think that we should use DBMS Name instead.

> ```yaml
> ###> doctrine/doctrine-bundle ###
>   database:
>     image: 'mysql:8.0'
>     environment:
>       - MYSQL_ROOT_PASSWORD=r007_p4ssw0rd
>       - MYSQL_DATABASE=databasename
>       - MYSQL_USER=mysqluser
>       - MYSQL_PASSWORD=p4ssw0rd
>     volumes:
>       - database_data:/var/lib/mysql:rw
> ###< doctrine/doctrine-bundle ###
> ```

Reviewer 3: ```suggestion
```diff
###> doctrine/doctrine-bundle ###
-   image: postgres:${POSTGRES_VERSION:-15}-alpine
+   image: mysql:${MYSQL_VERSION:-8}
    environment:
-     POSTGRES_DB: ${POSTGRES_DB:-app}
+     MYSQL_DATABASE: ${MYSQL_DATABASE:-app}
      # You should definitely change the password in production
+     MYSQL_ROOT_PASSWORD: ${MYSQL_ROOT_PASSWORD:-!ChangeMe!}
-     POSTGRES_PASSWORD: ${POSTGRES_PASSWORD:-!ChangeMe!}
+     MYSQL_PASSWORD: ${MYSQL_PASSWORD:-!ChangeMe!}
-     POSTGRES_USER: ${POSTGRES_USER:-app}
+     MYSQL_USER: ${MYSQL_USER:-app}
    volumes:
-     - database_data:/var/lib/postgresql/data:rw
+     - database_data:/var/lib/mysql:rw
      # You may use a bind-mounted host directory instead, so that it is harder to accidentally remove the volume and lose all your data!
-     # - ./docker/db/data:/var/lib/postgresql/data:rw
+     # - ./docker/db/data:/var/lib/mysql:rw
###< doctrine/doctrine-bundle ###
```
```

Depending on the database configuration, modify the environment in the same file at `services.php.environment.DATABASE_URL`
```

> DATABASE_URL: mysql://mysqluser:p4ssw0rd@database:3306/nucast?serverVersion=8.0.32&charset=utf8mb4

Reviewer 3: ```suggestion
DATABASE_URL: mysql://${MYSQL_USER:-app}:${MYSQL_PASSWORD:-!ChangeMe!}@database:3306/${MYSQL_DATABASE:-app}?serverVersion=${MYSQL_VERSION:-8}&charset=${MYSQL_CHARSET:-utf8mb4}
```

```

Since we changed the port, we also have to define this in the `compose.override.yaml`:

> ```yaml
> ###> doctrine/doctrine-bundle ###
>   database:
>     ports:
>       - "3306:3306"
> ###< doctrine/doctrine-bundle ###
> ```

Reviewer 3: ```suggestion
```diff
###> doctrine/doctrine-bundle ###
  database:
    ports:
-     - "5432:5432"
+     - "3306:3306"
###< doctrine/doctrine-bundle ###
```
```

> Last but not least, we need to install the mysql driver in `Dockerfile`:

Reviewer 3: ```suggestion
Last but not least, we need to install the MySQL driver in `Dockerfile`:
```

```diff
###> doctrine/doctrine-bundle ###

> - RUN install-php-extensions pdo_postgres
> + RUN install-php-extensions pdo_mysql

Reviewer 3: ```suggestion
-RUN install-php-extensions pdo_pgsql
+RUN install-php-extensions pdo_mysql
```

###< doctrine/doctrine-bundle ###
```

> ## Change environment

Reviewer 2: ```suggestion
## Change Environment
```

```dotenv 

> DATABASE_URL="mysql://mysqluser:p4ssw0rd@database:3306/databasename?serverVersion=8.0.32&charset=utf8mb4"

Reviewer 3: ```suggestion
DATABASE_URL=mysql://${MYSQL_USER:-app}:${MYSQL_PASSWORD:-!ChangeMe!}@database:3306/${MYSQL_DATABASE:-app}?serverVersion=${MYSQL_VERSION:-8}&charset=${MYSQL_CHARSET:-utf8mb4}
```

# Using MySQL

The Docker configuration of this repository is extensible thanks to Flex recipes. By default, the recipe installs PostgreSQL.
If you prefer to work with MySQL, follow these steps:

> First, install the doctrine/orm pack as described: `docker compose exec php composer req symfony/orm-pack`

Reviewer 3: ```suggestion
First, install the `symfony/orm-pack` package as described: `composer require symfony/orm-pack`
```

Reviewer 3: Installing the `symfony/orm-pack` package should not be done inside a container

Reviewer: Normally, it should be inside the container, but in our case it can be a bit tricky.
These files are listed in the `.dockerignore` file but shared in the `compose.override.yaml` and at this stage we are more in dev mode where these files are shared.

Author: Maybe I'm misunderstanding you, but isn't it exactly right? Working with composer outside the container is error-prone because the local PHP version does not necessarily match or other parameters are different (extensions, etc.)

Reviewer 3: The Symfony recipe applies to `compose.yml`, `compose.override.yaml`, `Dockerfile`, etc.

If the container is running in development mode, the changes will be applied in the local folder thanks to this line:
https://github.com/dunglas/symfony-docker/blob/f063b11fb8d3522e3811124f43b2145dad205199/compose.override.yaml#L8

Otherwise (production mode) it won't work, so I would recommend using this command outside the container to avoid any confusion. Anyway, this is just a recommendation. Perhaps we should use some uniform style for installing additional packages.

Reviewer 3: The Symfony recipe applies to `compose.yml`, `compose.override.yaml`, `Dockerfile`, etc.

If the container is running in development mode, the changes will be applied in the local folder:
https://github.com/dunglas/symfony-docker/blob/f063b11fb8d3522e3811124f43b2145dad205199/compose.override.yaml#L8

Otherwise (production mode) it won't work, so I would recommend using this command outside the container to avoid any confusion. Anyway, this is just a recommendation. Perhaps we should use some uniform style for installing additional packages.

Reviewer: In production mode, it won't work and I'm not sure we'll end up in that this situation, so i'm in favour of doing it inside.
However, it deserves to be mentioned see : https://github.com/dunglas/symfony-docker/pull/571/files

Reviewer: You're right @dsentker. The point raised by @bkosun is the files modified by recipes are not sync in the container in prod mode. So the the recipe can't be applied.

Author: > However, you raise a good point that deserves to be mentioned see : https://github.com/dunglas/symfony-docker/pull/571/files

Maybe add a second sentence like 

> [!NOTE] Be sure that your host machine does provide  the same requirements as defined in the composer configuration files.

## Docker Configuration
Change the database image to use MySQL instead of PostgreSQL in `compose.yaml`:

```diff
###> doctrine/doctrine-bundle ###
-   image: postgres:${POSTGRES_VERSION:-15}-alpine
+   image: mysql:${MYSQL_VERSION:-8}
    environment:
-     POSTGRES_DB: ${POSTGRES_DB:-app}
+     MYSQL_DATABASE: ${MYSQL_DATABASE:-app}
      # You should definitely change the password in production

> +     MYSQL_ROOT_PASSWORD: ${MYSQL_ROOT_PASSWORD:-!ChangeMe!}

Reviewer: ```suggestion
+     MYSQL_RANDOM_ROOT_PASSWORD: "true"
```

-     POSTGRES_PASSWORD: ${POSTGRES_PASSWORD:-!ChangeMe!}
+     MYSQL_PASSWORD: ${MYSQL_PASSWORD:-!ChangeMe!}
-     POSTGRES_USER: ${POSTGRES_USER:-app}
+     MYSQL_USER: ${MYSQL_USER:-app}
    volumes:
-     - database_data:/var/lib/postgresql/data:rw
+     - database_data:/var/lib/mysql:rw
      # You may use a bind-mounted host directory instead, so that it is harder to accidentally remove the volume and lose all your data!
-     # - ./docker/db/data:/var/lib/postgresql/data:rw
+     # - ./docker/db/data:/var/lib/mysql:rw
###< doctrine/doctrine-bundle ###
```

Depending on the database configuration, modify the environment in the same file at `services.php.environment.DATABASE_URL`
```
DATABASE_URL: mysql://${MYSQL_USER:-app}:${MYSQL_PASSWORD:-!ChangeMe!}@database:3306/${MYSQL_DATABASE:-app}?serverVersion=${MYSQL_VERSION:-8}&charset=${MYSQL_CHARSET:-utf8mb4}
```

Since we changed the port, we also have to define this in the `compose.override.yaml`:
```diff
###> doctrine/doctrine-bundle ###
  database:
    ports:

> -     - "5432:5432"
> +     - "3306:3306"

Reviewer 3: ```suggestion
-     - "5432"
+     - "3306"
```

###< doctrine/doctrine-bundle ###
```

Last but not least, we need to install the MySQL driver in `Dockerfile`:
```diff
###> doctrine/doctrine-bundle ###
-RUN install-php-extensions pdo_pgsql
+RUN install-php-extensions pdo_mysql
###< doctrine/doctrine-bundle ###
```

> ## Change Environment

Reviewer: ```suggestion
## Change Environment
Change the database configuration in `.env`:

```
